### DISCUSSION TITLE
Enable parameter estimation as long as all comparison levels are observed

### DISCUSSION THREAD

**Discussion by lucasmalherbe** [moj-analytical-services/splink#1613] [General]
Currently, the `estimate_parameters_using_expectation_maximisation` method disables estimation of parameters for all the fields used in the `blocking_rule` parameter.

This behaviour makes perfect sense for a blocking rule like `l.surname = r.surname`. It is impossible to get parameter estimates for the `surname `field here since all the remaining pairs agree on this field.

But with a blocking rule like `l.postcode = r.postcode OR left(l.date_of_birth, 4) = left(r.date_of_birth, 4)` which keeps all pairs agreeing either on the postcode or on the year of birth, it is theoretically possible to compute parameter estimates for all the fields. However, the method will disable estimation for the `postcode` and `date_of_birth` fields.

The documentation suggests to proceed in several steps, e.g. first use l.postcode = l.postcode as blocking rule to estimate all m probabilities except for `postcode`, and then use the `left(l.date_of_birth, 4) = left(r.date_of_birth, 4)` blocking rule to do another round of estimation. **Does this multi-step procedure create bias?**

Let’s say I have a field called `first_name` in my dataset. For this field, I will get an estimate with the first call to the `estimate_parameters_using_expectation_maximisation` method and then I will get a (probably slightly) different estimate with the second call because a different set of pairs is used. It looks like a source of bias. In particular, if I do the two rounds of estimation in reverse order, the final estimates will probably be different, and I do not really have a way to choose which one is the right order. Maybe that’s not significant though.

**Would it make sense, from a methodological standpoint, to proceed to the estimation in a single step, given that all comparison levels are observed in the set of pairs remaining after applying the blocking rule?**

In terms of code, I'm wondering if it would make sense to go from this :

```
training_blocking_rule = "l.postcode = r.postcode"
linker.estimate_parameters_using_expectation_maximisation(training_blocking_rule)
training_blocking_rule = "LEFT(l.date_of_birth, 4) = LEFT(r.date_of_birth, 4)"
linker.estimate_parameters_using_expectation_maximisation(training_blocking_rule)
```

to this :

```
training_blocking_rule = "l.postcode = r.postcode or left(l.date_of_birth, 4) = left(r.date_of_birth, 4)"
linker.estimate_parameters_using_expectation_maximisation(training_blocking_rule)
```

**Comment by samnlindsay** [comment]
Hi Lucas, I'll have a go at explaining why I don't think your preference for a single training blocking rule would work.

When blocking on **postcode**, _all comparisons have the same prior_ dependent only on the global "lambda" parameter (`linker._settings_obj._probability_two_random_records_match`) and the m and u probabilities of a match on postcode. These parameters are fixed, and all other model parameters can be estimated using EM. 

When blocking on **year of birth**, _all comparisons have the same prior_. All parameters can be estimated except the global lambda, and any m and u parameters dependent on year of birth.

When blocking on `l.postcode = r.postcode or left(l.date_of_birth, 4) = left(r.date_of_birth, 4)` each comparison has one of _at least three different priors_:
1. Prob(match | match on postcode AND year of birth)
2. Prob(match | match on postcode AND NOT year of birth)
3. Prob(match | match on year of birth AND NOT postcode)

Training a single model using EM requires enforcing a single prior probability for all comparisons and estimating the other parameters under the condition of that prior. **This model cannot be compatible with all 3 of the above.** 

The only way to properly "correct" for these distinct subsets is to already have trained model parameters for both columns beforehand. Therefore we must always have at least 2 training blocking rules so we can block on one (or several) columns to estimate parameters for the others, and then block on other columns to estimate the missing parameters from the first model.

Enforcing that these blocking rules be run separately actually _prevents_ bias by disentangling the selection criteria (and therefore the prior) from the parameters being estimated. Multiple training blocking rules run in a different order should have no significant effect given the parameters are (in principle) independent and the EM algorithm should iterate to the same solution regardless of small changes in starting values. This can be confirmed by using `linker.parameter_estimate_comparisons_chart()` to compare independent estimates using different training blocking rules.
